부산대 Android_김주송_5주차 과제 (STEP1)

jooiss opened the pull request:
## 코드 작성 시, 어려웠던 점
- 의존성 주입을 도입할 때 어느 부분까지 도입하는 것이 맞는지에 대한 고민이 많이 들었습니다.
- Hilt를 사용할 때, 제대로 사용하고 있는 건지에 대해 걱정이 되었습니다.

## 중점적으로 리뷰해주셨으면 하는 부분
- 제가 한 부분 외에도 어떤 부분에 의존성을 주입하는 것이 좋은지에 대해 알고 싶습니다. 
    (LiveData, DataBinding는 2단계에서 진행할 예정입니다.)
- 여러 Place 모델을 사용하던 것을 Room 사용할 때 만든 데이터 클래스(Place.kt)로 합쳤는데, 이를 data 패키지에 넣어야 할 지, domain 패키지에 넣어야 할 지 고민입니다. (데이터 형태를 나타낸다는 것에서는 domain 패키지가 맞을 것 같은데 Room을 적용할 때 사용해서 data 패키지에 더 잘 어울리는 게 아닌가 하는 고민이 들었습니다.)
- 패키지를 잘 분류했는지에 대해 확인받고 싶습니다.

항상 좋은 코멘트 남겨주셔서 감사합니다 멘토님 !😌

MyStoryG committed:
Initial commit

MyStoryG committed:
Initial commit

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Design search layout

jooiss committed:
feat: Add place database

- PlaceContract : 테이블 이름 및 열 이름 정의
- PlaceDatabaseHelper : 데이터베이스 생성
- PlaceDatabaseAccess : 데이터베이스에 데이터 저장 및 삭제 기능

jooiss committed:
fix: Add trigger to prevent storing duplicate data

jooiss committed:
feat: Design search recyclerView and place recyclerView

- place_item : place list item layout
- search_item : search list item layout

jooiss committed:
feat: Set the layout to be displayed

- PlaceDataModel : Place 데이터 클래스
- PlaceRecyclerViewAcitivity : Place 목록을 보여줄 recycler view 제어
- PlaceRecyclerViewAdapter : Place 목록 adapter

jooiss committed:
docs: Update README.md

- STEP2 구현 기능 목록 추가

jooiss committed:
feat: Add search function of place list

jooiss committed:
refactor: change place list item

jooiss committed:
refactor: resize search list item layout

jooiss committed:
feat: Add storing function of record list

MyStoryG committed:
Initial commit

jooiss committed:
fix: Resolve conflict

jooiss committed:
fix: Exclude kakao map dependency

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Add database insert function

jooiss committed:
fix: Resolve screen-off feature

jooiss committed:
feat: Set API key value for kakao API integration

jooiss committed:
feat: Add data transfer object file

jooiss committed:
feat: Add retrofit interface

jooiss committed:
refactor: Combine data activity and place activity

jooiss committed:
feat: Add retrofit object

jooiss committed:
feat: Update network security configuration

jooiss committed:
feat: Add database insertion function

jooiss committed:
refactor: Extract string to file

jooiss committed:
refactor: Change the search item design

jooiss committed:
feat: Reset the data at first

jooiss committed:
feat: Change to category search

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Add kakao native app key

jooiss committed:
feat: Add dependency about kakao map SDK

jooiss committed:
feat: Design map layout

jooiss committed:
feat: Add map view function using the kakao map API

jooiss committed:
fix: Resolve file redundancy

jooiss committed:
fix: set text to be written in one line

jooiss committed:
refactor: Change DTO file name

jooiss committed:
fix: Make API key resource not be saved

jooiss committed:
fix: Call MapView method in time for activity

jooiss committed:
fix: Change lifecycle function about kakao map API

jooiss committed:
fix: Change CategoryGroup finding method

jooiss committed:
refactor: Separate the Retrofit network-related function

jooiss committed:
fix: Resolve conflict

jooiss committed:
부산대 Android_김주송_4주차 과제 (STEP0) (#19)

* Initial commit

* docs: Update README.md

* feat: Design search layout

* feat: Add place database

- PlaceContract : 테이블 이름 및 열 이름 정의
- PlaceDatabaseHelper : 데이터베이스 생성
- PlaceDatabaseAccess : 데이터베이스에 데이터 저장 및 삭제 기능

* fix: Add trigger to prevent storing duplicate data

* feat: Design search recyclerView and place recyclerView

- place_item : place list item layout
- search_item : search list item layout

* feat: Set the layout to be displayed

- PlaceDataModel : Place 데이터 클래스
- PlaceRecyclerViewAcitivity : Place 목록을 보여줄 recycler view 제어
- PlaceRecyclerViewAdapter : Place 목록 adapter

* docs: Update README.md

- STEP2 구현 기능 목록 추가

* feat: Add search function of place list

* refactor: change place list item

* refactor: resize search list item layout

* feat: Add storing function of record list

* Initial commit

* fix: Exclude kakao map dependency

* docs: Update README.md

* feat: Add database insert function

* fix: Resolve screen-off feature

* feat: Set API key value for kakao API integration

* feat: Add data transfer object file

* feat: Add retrofit interface

* refactor: Combine data activity and place activity

* feat: Add retrofit object

* feat: Update network security configuration

* feat: Add database insertion function

* refactor: Extract string to file

* refactor: Change the search item design

* feat: Reset the data at first

* feat: Change to category search

* docs: Update README.md

* feat: Add kakao native app key

* feat: Add dependency about kakao map SDK

* feat: Design map layout

* feat: Add map view function using the kakao map API

* fix: Resolve file redundancy

* fix: set text to be written in one line

* refactor: Change DTO file name

* fix: Make API key resource not be saved

* fix: Call MapView method in time for activity

* fix: Change lifecycle function about kakao map API

* fix: Change CategoryGroup finding method

* refactor: Separate the Retrofit network-related function

---------

Co-authored-by: MyStoryG <bookwriterseo@gmail.com>

jooiss committed:
docs: Update README.md

jooiss committed:
refactor: Delete unnecessary files

jooiss committed:
feat: Change search method by keyword

jooiss committed:
feat: Display search result on selecting a saved search record

jooiss committed:
feat: Display modal when clicking on place list

jooiss committed:
feat: Show error message when onMapError

jooiss committed:
feat: Resume from last position on restart

jooiss committed:
test: Add UI test code for PlaceActivity

jooiss committed:
test: Add UI test code for MapActivity

jooiss committed:
test: Add unit test code for MapActivity

jooiss committed:
test: Add unit test code for PlaceActivity

jooiss committed:
Merge branch 'jooiss' into jooiss-step2

jooiss committed:
fix: Update UI test code for PlaceActivity

jooiss committed:
docs: Update README.md

jooiss committed:
부산대 Android_김주송_4주차 과제 (STEP1) (#71)

* Initial commit

* docs: Update README.md

* feat: Design search layout

* feat: Add place database

- PlaceContract : 테이블 이름 및 열 이름 정의
- PlaceDatabaseHelper : 데이터베이스 생성
- PlaceDatabaseAccess : 데이터베이스에 데이터 저장 및 삭제 기능

* fix: Add trigger to prevent storing duplicate data

* feat: Design search recyclerView and place recyclerView

- place_item : place list item layout
- search_item : search list item layout

* feat: Set the layout to be displayed

- PlaceDataModel : Place 데이터 클래스
- PlaceRecyclerViewAcitivity : Place 목록을 보여줄 recycler view 제어
- PlaceRecyclerViewAdapter : Place 목록 adapter

* docs: Update README.md

- STEP2 구현 기능 목록 추가

* feat: Add search function of place list

* refactor: change place list item

* refactor: resize search list item layout

* feat: Add storing function of record list

* Initial commit

* fix: Exclude kakao map dependency

* docs: Update README.md

* feat: Add database insert function

* fix: Resolve screen-off feature

* feat: Set API key value for kakao API integration

* feat: Add data transfer object file

* feat: Add retrofit interface

* refactor: Combine data activity and place activity

* feat: Add retrofit object

* feat: Update network security configuration

* feat: Add database insertion function

* refactor: Extract string to file

* refactor: Change the search item design

* feat: Reset the data at first

* feat: Change to category search

* docs: Update README.md

* feat: Add kakao native app key

* feat: Add dependency about kakao map SDK

* feat: Design map layout

* feat: Add map view function using the kakao map API

* fix: Resolve file redundancy

* fix: set text to be written in one line

* refactor: Change DTO file name

* fix: Make API key resource not be saved

* fix: Call MapView method in time for activity

* fix: Change lifecycle function about kakao map API

* fix: Change CategoryGroup finding method

* refactor: Separate the Retrofit network-related function

* docs: Update README.md

* refactor: Delete unnecessary files

* feat: Change search method by keyword

* feat: Display search result on selecting a saved search record

* feat: Display modal when clicking on place list

* feat: Show error message when onMapError

* feat: Resume from last position on restart

---------

Co-authored-by: MyStoryG <bookwriterseo@gmail.com>

LeeOhHyung committed:
Merge branch 'jooiss' into jooiss-step2

jooiss committed:
fix: Resolve merge conflict

jooiss committed:
refactor: Reorganize files according to data/domain/UI layer

jooiss committed:
chore: Rename application file name

jooiss committed:
refactor: Reorganize files according to clean architecture

jooiss committed:
refactor: Apply DataBinding in MapActivity and PlaceActivity

jooiss committed:
docs: Update README.md

jooiss committed:
feat: Add Place entity

jooiss committed:
feat: Add DAO interface

jooiss committed:
feat: Add Database

jooiss committed:
feat: Implement DBClient for managing Room database instance

jooiss committed:
refactor: Integrate Room database for data storage

jooiss committed:
test: Update test code for modified code

jooiss committed:
refactor: Add dependency injection for Retrofit

jooiss committed:
chore: Change Retrofit repository file name

jooiss committed:
refactor: Add dependency injection for Database

jooiss committed:
refactor: Add dependency injection for SharedPreferences

jooiss committed:
Merge branch 'jooiss' into jooiss-step1

jooiss committed:
fix: Resolve conflict

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/data/Place.kt:
```diff
@@ -0,0 +1,14 @@
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "places")
+data class Place(
+    @ColumnInfo val name: String,
+    @PrimaryKey val address: String,
+    @ColumnInfo val category: String?,
+    @ColumnInfo val x: String?,
+    @ColumnInfo val y: String?
+)
```
>여러 Place 모델을 사용하던 것을 Room 사용할 때 만든 데이터 클래스(Place.kt)로 합쳤는데, 이를 data 패키지에 넣어야 할 지, domain 패키지에 넣어야 할 지 고민입니다. (데이터 형태를 나타낸다는 것에서는 domain 패키지가 맞을 것 같은데 Room을 적용할 때 사용해서 data 패키지에 더 잘 어울리는 게 아닌가 하는 고민이 들었습니다.)

오,, 이 고민을 하셨군요. 실제로 고려해야할 부분이 맞습니다.
우선, 지금의 `Place` 클래스는 Room 에 저장하기 위한 목적이므로 data layer에서 사용되는 클래스 입니다. 하지만, 앱의 다른곳에서 사용할때도 똑같은 클래스를 사용하면 되는지 궁금하신것 같네요.

이런 Layer간에 사용하는 클래스가 달라지는 경우 실제로 2개를 만드는 편입니다. 예를들어 `PlaceEntity (data)` 와 `Place (domain)` 을 만들고, 서로간에 변환해주는 중간과정을 추가합니다. mapper 같은 녀석이 생겨나는거죠. 그리고 UI 에서는 domain 에 정의한 class를 사용하거나, UI에서 사용할 class를 별도로 만드는 방법이 있습니다.

```kotlin
// domain의 place -> data의 entity로 변환
fun PlaceEntity.toDomain(): Place {
  return Place(...)
}

// data의 entity -> domain의 place로 변환
fun Place.toEntity(): PlaceEntity {
  return PlaceEntity(...)
}
```

LeeOhHyung replied:
또하나 의견이 있습니다.

address를 PK로 지정하게 된다면, 모든 Place의 주소값이 중복되면 안된다는 전제조건이 붙습니다. 
만약, 주소가 동일한 Place가 DB에 삽입되면 정합성이 지켜지지 않을겁니다. 

- auto-generate 되는 id 필드를 만들고 pk로 사용하기 ([참고](https://tekken5953.tistory.com/57))
- address에 unique index를 걸어서 중복되지 않게 방지 ([참고](https://leveloper.tistory.com/161))
- dao insert 메소드에 `OnConflictStrategy` 적용하기  ([참고](https://coding-idiot.tistory.com/2))

제가 떠오르는 방법은 위 3가지가 있을 합니다. 적절하게 해결해보면 좋을것 같습니다.

jooiss replied:
첫번째 방법을 사용하여 uid를 만든 후, autoGenerate=true 설정을 하여 해결하려 하였습니다. 그러나 domain의 Place(name, address, category, x, y)와 data의 Place(uid, name, address, category, x, y)를 **mapping할 때 domain 영역에는 없는 uid 설정 부분**에서 해결책이 떠오르지 않아 다른 방법을 선택했습니다. 이 경우에 어떻게 대처해야 할까요? (0으로 기본 설정을 한 후, 하려했지만 그럴 경우 검색어 저장 DB에서 제대로 삭제 되지 않는 현상이 발생하였습니다.)

일단 지금은 primary key를 [name, address] 쌍으로 설정하고 onConflict 속성을 이용해 대체(REPLACE)하도록 설정해놨습니다.

jooiss replied:
2번 방안과 3번 방안을 함께 사용했습니다.

```PlaceEntity.kt
@Entity(tableName = "places", primaryKeys = ["name", "address"], indices = arrayOf(Index(value = ["name", "address"], unique = true)))
data class PlaceEntity(
    @ColumnInfo val name: String,
    @ColumnInfo val address: String,
    @ColumnInfo val category: String?,
    @ColumnInfo val x: String?,
    @ColumnInfo val y: String?
)
```

```PlaceDao.kt
@Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPlace(place: PlaceEntity)
```

LeeOhHyung replied:
@jooiss 
>mapping할 때 domain 영역에는 없는 uid 설정 부분에서 해결책이 떠오르지 않아 다른 방법을 선택했습니다. 이 경우에 어떻게 대처해야 할까요? 

아하, 이것도 실제로 많이 하는고민입니다. 저는 id필드도 domain model에 만드는 편입니다. 
그치만, 지금 적용하신 2개 컬럼을 pk로 하는것도 괜찮은 방법이네요

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/ui/SearchRecyclerViewAdapter.kt:
```diff
@@ -61,7 +61,7 @@ class SearchRecyclerViewAdapter(
         holder.placeName.text = place.name
     }
 
-    override fun submitList(list: MutableList<PlaceDataModel>?) {
+    override fun submitList(list: MutableList<Place>?) {
         super.submitList(list)
     }
```
`submitList` 를 override 하신 이유가 있을까요. 내부 구현을 보니 부모클래스의 메소드를 호출만 하고 있어서, 당장에는 제거해도 괜찮을것 같습니다

jooiss replied:
코드를 수정하는 과정에서 삭제해야 했는데 미처 발견하지 못했던 것 같습니다. 바로 삭제하였습니다!

LeeOhHyung commented on app/src/main/java/campus/tech/kakao/map/ui/MapActivity.kt:
```diff
@@ -82,15 +85,13 @@ class MapActivity : AppCompatActivity() {
-        val preferencesLatitude = preferences.getString("latitude", null)
-        val preferencesLongitude = preferences.getString("longitude", null)
+        val preferencesLatitude = sharedPreferences.getString("latitude", null)
+        val preferencesLongitude = sharedPreferences.getString("longitude", null)
         return Pair(preferencesLatitude, preferencesLongitude)
     }
 
     fun saveLocation() {
-        val preferences: SharedPreferences = getSharedPreferences("locationInfo", MODE_PRIVATE)
-        val editor: Editor = preferences.edit()
+        val editor: Editor = sharedPreferences.edit()
         editor.putString("latitude", latitude)
         editor.putString("longitude", longitude)
         editor.apply()
```
SharedPreference도 결국 기기에 특정 값들을 저장시키기 위한 목적이니, 데이터 삽입/조회는 data layer의 역할이라고 볼수있습니다. 따라서, PlaceRepository 혹은 위경도를 가져오는 역할의 Repository를 만들어서 사용해보는것이 어떨까요.

jooiss replied:
위경도를 가져오는 Module을 추가해봤습니다. 멘토님 의도대로 제가 맞게 한 건 지 조금 걱정이 됩니다만 STEP2에서 ViewModel을 결합하여 잘 적용시켜보겠습니다 감사합니다!

LeeOhHyung replied:
https://github.com/kakao-tech-campus-2nd-step2/android-map-refactoring/pull/27/commits/7a16ce0cc668009034641dee013cec3e204e8422 이 커밋 확인했습니다!

LeeOhHyung reviewed (commented):
@jooiss 5주차 과제 step1 하시느라 고생많으셨습니다! 코멘트 확인한번 부탁드려요

jooiss committed:
refactor: Separate Domain and Data layers for Place representation

jooiss committed:
feat: Handle duplicate entries to maintain data integrity

- Update schema to use [name, address] pair as unique identifier
- Implemente OnConflictStrategy for handling duplicates

jooiss committed:
refactor: Remove unnecessary code

jooiss committed:
feat: Add indices on 'name' and 'address' columns in PlaceEntity

jooiss committed:
refactor: Add dependency injection for latitude and longitude

LeeOhHyung reviewed (approved):
@jooiss  LGTM~! 코멘트 반영해서 커밋해주신거 확인했습니다. 머지하도록 할게요
